pkg/server/backend_manager.go

```diff
@@ -49,6 +49,25 @@ type Backend struct {
 	// cached from conn.Context()
 	id     string
 	idents header.Identifiers
+
+	// draining indicates if this backend is draining and should not accept new connections
+	draining bool
+	// mu protects draining field
```

Comment:
Consider using `sync/atomic` (e.g., `atomic.Bool` or `int32`) for the `draining` field. This would allow `IsDraining()` to be lock-free, reducing the overhead inside the O(N) loop in `GetRandomBackend`, which runs under the global `s.mu` lock.